### PR TITLE
1.15(Beta) Add tests for a Book functionality

### PR DESCRIPTION
**Pull request by OleksiiVitiuk**

### CODE REVIEW

**Comment by OleksiiVitiuk** [comment]
Now the tests look like this, please help find the bugs so that everything works. Thank you!
Screenshot_7

**Comment by OleksiiVitiuk** [comment]
image

**Comment by OleksiiVitiuk** [comment]
image

File: src/test/java/bookstore/TestUtil.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import java.math.BigDecimal;
+import java.util.Set;
+
+public class TestUtil {
+    private static final CreateBookRequestDto CREATE_BOOK_REQUEST_DTO
+            = new CreateBookRequestDto();
+    private static final CategoryCreateDto CATEGORY_CREATE_DTO
+            = new CategoryCreateDto();
+    private static final Category CATEGORY = new Category();
+    private static final Book BOOK = new Book();
```
Comment (JJJazl) [inline src/test/java/bookstore/TestUtil.java:16]: These are not constants, use lowercase letters

File: src/test/java/bookstore/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                .andReturn();
+
+        BookDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), BookDto.class);
+        assertBook(actual,
+                1L,
+                "BookTitle", "BookAuthor",
+                "123-456-789",
+                BigDecimal.valueOf(20.00),
+                "desc", "path");
```
Comment (JJJazl) [inline src/test/java/bookstore/controller/BookControllerTest.java:63]: Create book with expected values and compare actual book with expected. You can use reflectionEquals() method with assertTrue(). Fix everywhere

File: src/test/java/bookstore/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @Test
+    @DisplayName("Test get book by id")
+    void testGetBookById() throws Exception {
```
Comment (JJJazl) [inline src/test/java/bookstore/controller/BookControllerTest.java:51]: Let's change naming for all test methods
```suggestion
    void getBookById_ValidId_Ok() throws Exception {

```

File: src/test/java/bookstore/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+@SpringBootTest
+@AutoConfigureMockMvc
+@WithMockUser(username = "test", roles = {"USER", "ADMIN"})
+@Sql(scripts = {
+        "classpath:database/book/add-category-for-book.sql",
+        "classpath:database/book/add-book-to-table.sql"
+}, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+@Sql(scripts = "classpath:database/book/delete-book-and-category.sql",
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+class BookControllerTest {
```
Comment (JJJazl) [inline src/test/java/bookstore/controller/BookControllerTest.java:41]: Add tests for create/delete + negative test cases

File: src/test/java/bookstore/controller/CategoryControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+@SpringBootTest
+@AutoConfigureMockMvc
+@WithMockUser(username = "test", roles = {"USER", "ADMIN"})
+@Sql(scripts = {
+        "classpath:database/book/add-category-for-book.sql",
+        "classpath:database/book/add-book-to-table.sql"
+}, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+@Sql(scripts = "classpath:database/book/delete-book-and-category.sql",
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+public class CategoryControllerTest {
```
Comment (JJJazl) [inline src/test/java/bookstore/controller/CategoryControllerTest.java:38]: Same comments

File: src/test/java/bookstore/repository/BookRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        assertBook(TestUtil.getBook(), actualBooks.getContent().get(0));
+    }
+
+    private void assertBook(Book expected, Book actual) {
+        assertNotNull(actual);
+        assertEquals(expected.getId(), actual.getId());
+        assertEquals(expected.getTitle(), actual.getTitle());
+        assertEquals(expected.getAuthor(), actual.getAuthor());
+        assertEquals(expected.getIsbn(), actual.getIsbn());
+        assertEquals(0, expected.getPrice().compareTo(actual.getPrice())); // Для BigDecimal
```
Comment (JJJazl) [inline src/test/java/bookstore/repository/BookRepositoryTest.java:74]:
```suggestion
        assertEquals(0, expected.getPrice().compareTo(actual.getPrice()));

```

File: src/test/java/bookstore/service/BookServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+
+@ExtendWith(MockitoExtension.class)
+class BookServiceTest {
```
Comment (JJJazl) [inline src/test/java/bookstore/service/BookServiceTest.java:30]: Add negative test cases

File: src/test/java/bookstore/service/CategoryServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+
+@ExtendWith(MockitoExtension.class)
+class CategoryServiceImplTest {
```
Comment (JJJazl) [inline src/test/java/bookstore/service/CategoryServiceImplTest.java:27]: Same comments

**Review by JJJazl — CHANGES_REQUESTED** [review]
Good! Let's improve your solution

File: src/main/java/bookstore/controller/BookController.java [hunk 9]
```diff
@@ -72,8 +73,11 @@ public void deleteBook(@PathVariable Long id) {
             description = "Returns a list of books that match the search criteria")
     @GetMapping("/search")
     @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
-    public Page<BookDto> searchBooks(BookSearchParametersDto searchParameters,
+    public Page<BookDto> searchBooks(@RequestParam(required = false) String title,
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/BookController.java:76]: revert changes

File: src/main/java/bookstore/entity/Category.java [hunk 10]
```diff
@@ -13,8 +13,8 @@
 import org.hibernate.annotations.SQLRestriction;
 
 @Entity
-@Table(name = "categories")
-@SQLDelete(sql = "UPDATE category SET is_deleted = 1 WHERE id = ?")
+@Table(name = "category")
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/entity/Category.java:16]: table name should be plural
```suggestion
@Table(name = "categories")
```

File: src/main/resources/db/changelog/changes/08-create-category-table.yaml [hunk 11]
```diff
@@ -4,7 +4,7 @@ databaseChangeLog:
       author: oleksii
       changes:
         - createTable:
-            tableName: categories
+            tableName: category
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-create-category-table.yaml:7]: revert

File: src/test/java/bookstore/TestUtil.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
+package bookstore;
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/TestUtil.java:1]:
```suggestion
package bookstore.util;
```

File: src/test/java/bookstore/controller/CategoryControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+                = objectMapper.readValue(contentNode.toString(), new TypeReference<>() {});
+
+        assertNotNull(books);
+        assertEquals(1, books.size());
+
+        BookDtoWithoutCategoryIds actualBookWithoutCategory = books.get(0);
+        assertBookWithoutCategory(actualBookWithoutCategory);
+    }
+
+    private void assertBookWithoutCategory(BookDtoWithoutCategoryIds book) {
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/controller/CategoryControllerTest.java:112]: private method should be placed after public methods

File: src/test/java/bookstore/controller/CategoryControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+                        .content(json)
+                        .with(csrf())
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isCreated())
+                .andReturn();
+
+        CategoryDto actual = objectMapper
+                .readValue(result.getResponse().getContentAsString(), CategoryDto.class);
+        assertNotNull(actual.getId());
+        assertEquals(categoryDto.getName(), actual.getName());
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/controller/CategoryControllerTest.java:87]: you can just compare actual and expected

File: src/test/java/bookstore/controller/CategoryControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        JsonNode contentNode = objectMapper.readTree(result.getResponse().getContentAsString()).get("content");
+        List<CategoryDto> categories = objectMapper.readValue(contentNode.toString(), new TypeReference<>() {});
+
+        assertNotNull(categories);
+        assertEquals(categories.size(), 1);
+        assertEquals("Fantasy", categories.get(0).getName());
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/controller/CategoryControllerTest.java:60]: Can you compare content?

File: src/test/java/bookstore/controller/CategoryControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        assertEquals("BookAuthor", book.getAuthor());
+        assertEquals("123-456-789", book.getIsbn());
+        assertEquals(BigDecimal.valueOf(20.00), book.getPrice());
+        assertEquals("desc", book.getDescription());
+        assertEquals("path", book.getCoverImage());
+    }
+
+    @Test
+    @DisplayName("Test not valid category to create")
+    void validCategoryTest() throws Exception {
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/controller/CategoryControllerTest.java:125]: rename method, not clear what you are checking

File: src/test/java/bookstore/service/BookServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        assertEquals(expectedBookDto, actualBookDto);
+        verify(bookRepository).findById(book.getId());
+        verify(bookRepository).save(book);
+        verify(bookMapper).toDto(updatedBook);
+    }
+
+    @Test
+    @DisplayName("Search books by title")
+    void searchBooks_ShouldReturnFilteredBooks() {
+        // Big given
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/service/BookServiceTest.java:86]: remove comments or add them everywhere,  be consistent

File: src/test/java/bookstore/service/CategoryServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        verify(categoryRepository).findAll(pageable);
+    }
+
+    @Test
+    @DisplayName("Save category when save operation fails")
+    void saveCategory_ShouldThrowException_WhenSaveFails() {
+        when(categoryMapper.toModel(categoryCreateDto)).thenReturn(category);
+        when(categoryRepository.save(category)).thenThrow(new RuntimeException("Database error"));
+
+        try {
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/service/CategoryServiceImplTest.java:134]: don't use try catch in tests

Comment (Elena-Bruyako) [reply src/test/java/bookstore/service/CategoryServiceImplTest.java:134]: you can use assertThrows

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/test/java/bookstore/service/CategoryServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+        } catch (RuntimeException e) {
+            assertEquals("Category is not found with id: 1", e.getMessage());
+        }
+
+        verify(categoryRepository).findById(1L);
+    }
+
+    @Test
+    @DisplayName("Save category with invalid data")
+    void saveCategory_ShouldThrowException_WhenDataIsInvalid() {
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/service/CategoryServiceImplTest.java:159]: don't need to test validation

File: src/test/java/bookstore/service/CategoryServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+                () -> categoryService.saveCategory(categoryCreateDto));
+
+        assertEquals("Database error", exception.getMessage());
+
+        verify(categoryRepository).save(category);
+    }
+
+    @Test
+    @DisplayName("Update category when category does not exist")
+    void updateCategory_ShouldThrowException_WhenCategoryNotFound() {
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/service/CategoryServiceImplTest.java:145]: when category not exists in DB EntityNotFoundException should be thrown 
you can check it by `assertThrows(EntityNotFoundException.class, () -> ...`

File: src/test/java/bookstore/service/CategoryServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+        assertNotNull(result);
+        assertEquals(0, result.getTotalElements());
+        assertEquals(0, result.getContent().size());
+
+        verify(categoryRepository).findAll(pageable);
+    }
+
+    @Test
+    @DisplayName("Save category when save operation fails")
+    void saveCategory_ShouldThrowException_WhenSaveFails() {
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/service/CategoryServiceImplTest.java:131]: remove this test case

File: src/test/java/bookstore/service/BookServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    @Test
+    @DisplayName("Update book - exception(not found)")
+    void updateBook_ShouldThrowException_WhenBookNotFound() {
+        Long bookId = 100L;
+        CreateBookRequestDto requestDto = TestUtil.getRequestToUpdateBook();
+        when(bookRepository.findById(bookId)).thenReturn(Optional.empty());
+
+        RuntimeException exception =
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/service/BookServiceTest.java:132]: there is should be EntitNotFoundException

File: src/test/java/bookstore/service/BookServiceTest.java [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+        verify(bookMapper).toDto(book);
+    }
+
+    @Test
+    @DisplayName("Get book by id - exception(not found)")
+    void getBookById_ShouldThrowException_WhenNotFound() {
+        Long bookId = 99L;
+        when(bookRepository.findById(bookId)).thenReturn(Optional.empty());
+
+        RuntimeException exception =
```
Comment (Elena-Bruyako) [inline src/test/java/bookstore/service/BookServiceTest.java:116]: same here

File: src/test/java/bookstore/service/CategoryServiceImplTest.java (hunk 19 above)
Comment (JJJazl) [reply src/test/java/bookstore/service/CategoryServiceImplTest.java:147]: Not fixed, just delete this test

**Review by JJJazl — APPROVED** [review]
Great! Left one minor comment. Please pay attention to it and fix before moving on to the next task
